File: problems/Python/37. Sudoku Solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxs = [set() for _ in range(9)]
        
        empt = []
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':
                    num = board[i][j]
                    rows[i].add(num)
                    cols[j].add(num)
                    boxs[i//3*3+j//3].add(num)
                else:
                    empt.append((i, j))
        
        def back(idx):
            if idx == len(empt):
                return True
            i, j = empt[idx]
            box = i//3*3+j//3
            for num in '123456789':
                if num not in rows[i] and num not in cols[j] and num not in boxs[box]:
                    board[i][j] = num
                    rows[i].add(num)
                    cols[j].add(num)
                    boxs[box].add(num)
                    if back(idx+1):
                        return True
                    rows[i].remove(num)
                    cols[j].remove(num)
                    boxs[box].remove(num)
                    board[i][j] = '.'
            return False
        
        back(0)

File: problems/C++/37. Sudoku Solver.cpp
class Solution {
public:
    bool rows[9][9] = {};
    bool cols[9][9] = {};
    bool boxs[9][9] = {};
    
    void solveSudoku(vector<vector<char>>& board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '1';
                    rows[i][num] = true;
                    cols[j][num] = true;
                    boxs[i/3*3+j/3][num] = true;
                }
            }
        }
        back(board, 0, 0);
    }
    
    bool back(vector<vector<char>>& board, int i, int j) {
        if (i == 9) return true;
        if (j == 9) return back(board, i+1, 0);
        if (board[i][j] != '.') return back(board, i, j+1);
        
        int box = i/3*3+j/3;
        for (int num = 0; num < 9; num++) {
            if (!rows[i][num] && !cols[j][num] && !boxs[box][num]) {
                board[i][j] = num + '1';
                rows[i][num] = cols[j][num] = boxs[box][num] = true;
                if (back(board, i, j+1)) return true;
                rows[i][num] = cols[j][num] = boxs[box][num] = false;
                board[i][j] = '.';
            }
        }
        return false;
    }
};
